File: AEDs/Algoritmos de Ordenacao/C/Insertion.c
#include "Geracao.h"
#include "Log.h"

void sort ( int* arranjo, int tamanho, Log* log )
{
    for( int i = 1; i < tamanho; i = i + 1 ) 
    {
        int tmp = arranjo[i];
        int j = i - 1;
        incrementarMovimentacoes( log, 1 );
        
        incrementarComparacoes( log, 2 );
        while( (j >= 0) && (arranjo[j] > tmp) ) 
        {
            arranjo[j + 1] = arranjo[j];
            incrementarMovimentacoes( log, 1 );
            j--;
        } // end while
        arranjo[j+1] = tmp;
        incrementarMovimentacoes( log, 1 );
    } // end for
} // end sort ( )

int main ( void )
{
    Geracao* arranjo = new_Geracao( 100 );
    Log*     log     = new_Log ( );

    printf( "\n//----- INSERTION SORT -----//\n");

    preencher_aleatorio( arranjo, 200 );

    printf("\nARRANJO PRE ORDENACAO:\n");
    mostrar( arranjo );

    inicioTempo( log );
    sort( arranjo->arranjo, arranjo->tamanho, log );
    fimTempo( log );

    printf("\nARRANJO POS ORDENACAO:\n");
    mostrar( arranjo );

    registrar ( log );

    delete_Log ( log );
    delete_Geracao( arranjo );
} // end main ( )

File: AEDs/Algoritmos de Ordenacao/Java/Insertion.java
public class Insertion extends Geracao 
{
    public Insertion ( )
    {
        super ( );
    }

    public Insertion ( int tamanho )
    {
        super( tamanho );
    }

    /**
     *  <p>Insertion Sort é um algoritmo de ordenação que constrói uma lista ordenada
     *  um item por vez, movendo cada novo item para a posição correta na lista
     *  ordenada.</p> 
     *  
     *  <p>Complexidade: O(n^2) no pior caso, O(n) no melhor caso. </p>
     *  <p>Número de comparações: O(n^2).                          </p>
     *  <p>Número de movimentações: O(n^2).                        </p>
     */
    @Override
    public void sort( Log log )
    {
        for( int i = 1; i < tamanho; i = i + 1 ) 
        {
            int key = arranjo[i];
            int j = i - 1;
            log.incrementarMovimentacoes();

            log.incrementarComparacoes( 2 );
            while( ( j >= 0 ) && ( arranjo[j] > key ) ) 
            {
                arranjo[j+1] = arranjo[j]; // deslocamento
                log.incrementarMovimentacoes();
                j--;
            } // end while
            arranjo[j+1] = key;
            log.incrementarMovimentacoes();
        } // end for
    } // end sort ( )

    public static void main( String[] args ) 
    {
        Insertion arranjo = new Insertion( 100 );
        Log log = new Log();

        System.out.println( "\n//----- INSERTION SORT -----//");

        arranjo.preencher_aleatorio();

        System.out.println("\nARRANJO PRE ORDENACAO:");
        arranjo.mostrar( );

        log.inicioTempo();
        arranjo.sort( log );
        log.fimTempo();

        System.out.println("\nARRANJO POS ORDENACAO:");
        arranjo.mostrar( );

        log.mostrar();
    } // end main ( )
}
